File: GaniOSApp/GanClient/GanClient/Binary/BinaryWritable.swift
//
//  BinaryWritable.swift
//  GanClient
//
//  Created by tao on 2016/4/1.
//  Copyright © 2016年 Xattacker. All rights reserved.
//

import Foundation


// fluent interface
public protocol BinaryWritable
{
    associatedtype Writable: BinaryWritable
    
    func writeData(_ data: Data) -> Writable
    func writeBinary(_ bytes: UnsafeRawPointer, offset: UInt, length: UInt) -> Writable
    func writeShort(_ value: Int16) -> Writable
    func writeInteger(_ value: Int32) -> Writable
    func writeLongLong(_ value: Int64) -> Writable
    func writeDouble(_ value: Double) -> Writable
    func writeString(_ value: String) -> Writable
}

File: GanAndroidApp/GanClient/src/main/java/com/xattacker/binary/BinaryWritable.kt
package com.xattacker.binary

import java.io.IOException

// fluent interface
interface BinaryWritable<T>
{
    @Throws(IOException::class)
    fun writeBinary(data: ByteArray, offset: Int, length: Int): T

    @Throws(IOException::class)
    fun writeShort(short: Short): T

    @Throws(IOException::class)
    fun writeInteger(int: Int): T

    @Throws(IOException::class)
    fun writeLong(long: Long): T

    @Throws(IOException::class)
    fun writeDouble(double: Double): T

    @Throws(IOException::class)
    fun writeString(str: String): T
}
